onnxruntime/core/session/inference_session.cc

```diff
@@ -415,27 +415,28 @@ class InferenceSession::Impl {
                   "Unexpected input data type. Actual: (" + actual_name + ") , expected: (" + expected_name + ")");
   }
 
-  common::Status ValidateInputTypes(const std::vector<std::string>& feed_names,
-                                    const std::vector<MLValue>& feeds) {
+  common::Status ValidateInputs(const std::vector<std::string>& feed_names,
+                                const std::vector<MLValue>& feeds) {
     const auto begin_names = feed_names.cbegin();
     const auto end_names = feed_names.cend();
-    for (auto& arg : input_def_list_) {
+    for (auto& arg : required_input_def_list_) {
```

Comment:
>required_input_def_list_ [](start = 21, length = 24)

I don't think you can just arbitrarily ignore checking types of optional inputs.

I'd suggest 
 1) Check all required inputs are present
 2) if feed_names.size() == required_input_def_list_.size() you could just check types for the required inputs, otherwise you need to fallback to iterating the full list. AFAIK passing optional inputs is exceedingly rare, so that should achieve essentially the same performance win in a safer manner.

Reply:
Thank you very much! As discussed, we don't allow overwriting weights in inference cases. Meanwhile, I added the logic to ensure that caller provides exactly the inputs needed. please kindly take one more look.

Reply:
I think it may be better to iterate over the input vector (feeds) and do a lookup on required_input_def_list_ (which I hope will be a hash-lookup). Won't this avoid the quadratic behavior also?

Reply:
Would need something extra in that case to handle if there were duplicate names in the feeds. If you iterate the required names you guarantee they're all there. 

---
In reply to: [261732911](https://github.com/Microsoft/onnxruntime/pull/532#discussion_r261732911) [](ancestors = 261732911)

Reply:
True. You'd need a set to track and identify duplicates. We could add that (though it is already present as a dictionary up the callstack!) The earlier code didn't require a nested loop since we had dictionaries/hash-sets on both sides, allowing a constant-time loop. While the difference may not matter often, this does involve equality comparison of strings. If the model uses multiple inputs with names that have a long common prefix, it could be noticeable. Or, may be not much.